book/visual-effects.md

```diff
@@ -648,338 +631,309 @@ of thumb is: if you don't need a non-default blend mode, then you can use
-know what is actually going on in these simple-looking Skia APIs. We'll write
-some Python functions[^pedagogical] that do the math of these operations,
-starting with `apply_opacity`, which simply multiplies the alpha channel by the
-given opacity.[^see-alpha]
+Now, _before_ we add our temporary command list to the overall display
+list, we can use `SaveLayer` to add transparency to the whole element.
+I'm going to do this in a new `paint_visual_effects` method, because
+we'll want to make the same changes to all of our other layout
+objects:
 
-[^see-alpha]: Now might be a good time to go back and re-read the paragraph
-about alpha [here](#alpha).
-
-[^pedagogical]: These Python functions will not end up as part of your browser.
-We're writing them down only so that it's clear what exactly Skia is doing.
+``` {.python}
+class BlockLayout:
+    def paint(self, display_list):
+        # ...
+        cmds = paint_visual_effects(self.node, cmds, rect)
+        display_list.extend(cmds)
+```
 
-Let's assume that pixels are represented in a `skia.Color4f`, which has three
-color properties `fR`, `fG`, and `fB` (floating-point red/green/blue, between 0
-and 1), and one alpha property `fA` (floating-point alpha).[^alpha-vs-opacity]
+Inside `paint_visual_effects`, we'll parse the opacity value and
+construct the appropriate `SaveLayer`:
 
-[^alpha-vs-opacity]: The difference between opacity and alpha is often
-confusing. To remember the difference, think of opacity as a visual effect
-*applied to* content, but alpha as a *part of* content. In fact, whether there
-is an alpha channel in a color representation at all is often an implementation
-choice---sometimes graphics libraries instead multiply the other color channels
-by the alpha amount, which is called a *premultiplied* representation of the color.
+``` {.python}
+def paint_visual_effects(node, cmds, rect):
+    opacity = float(node.style.get("opacity", "1.0"))
 
-``` {.python.example}
-# Returns |color| with opacity applied. |color| is a skia.Color4f.
-def apply_opacity(color, opacity):
-    new_color = color
-    new_color.fA = color.fA * opacity
-    return new_color
+    return [
+         SaveLayer(skia.Paint(Alphaf=opacity), cmds),
+    ]
 ```
 
-Next, compositing: let's also assume that the coordinate spaces and pixel
-densities of the two surfaces are the same, and therefore their pixels overlap
-and have a 1:1 relationship. Therefore we can "composite" each pixel in the
-popped surface on top of (into, really) its corresponding pixel in the restored
-surface. Let's call the popped surface the *source surface* and the restored
-surface the *backdrop surface*. Likewise each pixel in the popped surface is
-a *source pixel* and each pixel in the restored surface is a *backdrop pixel*.
+Note that `paint_visual_effects` receives a list of commands and
+returns another list of commands. It's just that the output list is
+always a single `SaveLayer` command.
```

Comment:
```suggestion
always a single `SaveLayer` command that wraps the original content---which makes sense, because
first we need to draw the commands to a surface, and *then* apply transparency to it when blending into
the parent.
```